xarray/namedarray/core.py

```diff
@@ -0,0 +1,337 @@
+from __future__ import annotations
+
+import copy
+import itertools
+import math
+import typing
+from collections.abc import Hashable, Iterable
+from dataclasses import dataclass, field
+
+import numpy as np
+import numpy.typing as npt
+
+from xarray.namedarray.utils import Frozen, _default, is_duck_dask_array
+
+# temporary placeholder for indicating that an array api compliant
+# type. hopefully in the future we can narrow this down more
+T_Array = typing.TypeVar("T_Array", bound=typing.Any)
+
+
+@dataclass(slots=True)
+class NamedArray:
+    dims: tuple[Hashable, ...] = field(init=True)
+    data: npt.ArrayLike = field(init=True)
+    attrs: dict[typing.Any, typing.Any] = field(default_factory=dict, init=True)
+
+    _dims: tuple[Hashable, ...] = field(init=False)
+    _data: npt.ArrayLike = field(init=False)
+    _attrs: dict[typing.Any, typing.Any] = field(default_factory=dict, init=False)
+
```

Comment:
- This doesn't work in Python 3.9 because `slots` was introduced version 3.10.

```File "/home/runner/work/xarray/xarray/xarray/__init__.py", line 1, in <module>
    from xarray import testing, tutorial
  File "/home/runner/work/xarray/xarray/xarray/testing.py", line 11, in <module>
    from xarray.core.coordinates import Coordinates
  File "/home/runner/work/xarray/xarray/xarray/core/coordinates.py", line [17](https://github.com/pydata/xarray/actions/runs/5932971832/job/16087703883?pr=8075#step:8:18), in <module>
    from xarray.core.alignment import Aligner
  File "/home/runner/work/xarray/xarray/xarray/core/alignment.py", line 24, in <module>
    from xarray.core.variable import Variable, as_compatible_data, calculate_dimensions
  File "/home/runner/work/xarray/xarray/xarray/core/variable.py", line 48, in <module>
    from xarray.namedarray.core import NamedArray
  File "/home/runner/work/xarray/xarray/xarray/namedarray/core.py", line [20](https://github.com/pydata/xarray/actions/runs/5932971832/job/16087703883?pr=8075#step:8:21), in <module>
    @dataclass(slots=True)
TypeError: dataclass() got an unexpected keyword argument 'slots'
```

- also, i'm not sure of a proper way to maintain the usage of private attributes (`._data`, and `._dims`, etc) while using `.data`, `.dims`, etc fields. 

```python
class NamedArray:
    dims: tuple[Hashable, ...] = field(init=True)
    data: npt.ArrayLike = field(init=True)
    attrs: dict[typing.Any, typing.Any] = field(default_factory=dict, init=True)

    _dims: tuple[Hashable, ...] = field(init=False)
    _data: npt.ArrayLike = field(init=False)
    _attrs: dict[typing.Any, typing.Any] = field(default_factory=dict, init=False)
```

@Illviljan, after experimenting with `dataclass`, i am contemplating reverting the changes. initially, i thought it would be a simple process, but i encountered inheritance issues that would require significant adjustments to make everything function properly. 

i believe it would be beneficial to revisit the usage of dataclasses in a separate pull request at a later time.

Reply:
Good try! 

If you want you could try the combining properties and dataclasses: https://florimond.dev/en/posts/2018/10/reconciling-dataclasses-and-properties-in-python/

Reply:
I agree with not using dataclasses here. Dataclasses is convenient for writing quick classes, but on NamedArray we will need custom versions of all the special methods created by the dataclass decroator anyways (`__init__`, `__repr__`, `__eq__`, etc). So there's really no advantage to using a dataclass.